### PR TITLE
Replication changes

### PR DESCRIPTION
**Pull request by NyakudyaA**
* New environment variables called REPLICATION_USER and REPLICATION_PASS
* Replication now uses a specific role with only replication permission
* Added gdal in the container
* Register gdal drivers with pg raster
* Support out-of-db raster

### CODE REVIEW

**Comment by NyakudyaA** [comment]
@gubuntu can you take a look at this PR

File: Dockerfile [hunk 1]
```diff
@@ -31,6 +31,13 @@ ADD setup.sh /setup.sh
 RUN chmod +x /setup.sh
 RUN /setup.sh
 
+ADD locale.gen /etc/locale.gen
+RUN /usr/sbin/locale-gen
+ENV LANG en_US.UTF-8
+ENV LANGUAGE en_US:en
+ENV LC_ALL en_US.UTF-8
+RUN update-locale ${LANG}
+
```
Comment (gubuntu) [inline Dockerfile:40]: aren't these all defaults? why do we need to add them explicitly? To give users a place to change them if they want?

File: README.md [hunk 2]
```diff
@@ -218,15 +220,23 @@ mirror database content from a designated master. This replication scheme allows
 us to sync databases. However a `replicant` is only for read-only transaction, thus 
 we can't write new data to it. The whole database cluster will be replicated.
 
+### Database permissions
+Since we are using a role ${REPLICATION_USER}, we need to ensure that he has access to all
```
Comment (gubuntu) [inline README.md]: 'it' not 'he'

File: README.md [hunk 3]
```diff
@@ -218,15 +220,23 @@ mirror database content from a designated master. This replication scheme allows
...
 we can't write new data to it. The whole database cluster will be replicated.
 
+### Database permissions
+Since we are using a role ${REPLICATION_USER}, we need to ensure that he has access to all 
+the tables in a particular schema. So if a user adds another schema called `data`
+to the database `gis` he also has to update the permission for the user
+with the following SQL assuming the ${REPLICATION_USER} is called replicator
+
+     ALTER DEFAULT PRIVILEGES IN SCHEMA data GRANT SELECT ON TABLES TO replicator;
+
```
Comment (gubuntu) [inline README.md:230]: why? The replication role is JUST for replication and not for using the slave. Normal users will use the slave. So the replication role won't be doing select statements.

File: setup-database.sh [hunk 4]
```diff
@@ -47,7 +47,10 @@ source /setup-user.sh
 # Create a default db called 'gis' or $POSTGRES_DBNAME that you can use to get up and running quickly
 # It will be owned by the docker db user
 # Since we now pass a comma separated list in database creation we need to search for all databases as a test
-
+touch custom.sql
+cat >> custom.sql <<EOF
+ALTER DEFAULT PRIVILEGES IN SCHEMA public GRANT SELECT ON TABLES TO ${REPLICATION_USER}
+EOF
```
Comment (gubuntu) [inline setup-database.sh:53]: as per previous comment, why is this necessary?

File: setup-database.sh [hunk 5]
```diff
@@ -60,11 +63,13 @@ for db in $(echo ${POSTGRES_DBNAME} | tr ',' ' '); do
             echo "Loading legacy sql"
             su - postgres -c "psql ${db} -f ${SQLDIR}/legacy_minimal.sql" || true
             su - postgres -c "psql ${db} -f ${SQLDIR}/legacy_gist.sql" || true
+            export PGPASSWORD=${POSTGRES_PASS}
+            psql ${db} -U ${POSTGRES_USER} -p 5432 -h localhost -f custom.sql
```
Comment (gubuntu) [inline setup-database.sh:67]: postgres user or replication user?

File: setup-user.sh [hunk 6]
```diff
@@ -19,9 +19,19 @@ echo "postgresql user: $POSTGRES_USER" > /tmp/PGPASSWORD.txt
...
 
 # Check user already exists
+echo "Creating superuser $POSTGRES_USER"
 RESULT=`su - postgres -c "psql postgres -t -c \"SELECT 1 FROM pg_roles WHERE rolname = '$POSTGRES_USER'\""`
 COMMAND="ALTER"
 if [ -z "$RESULT" ]; then
 	COMMAND="CREATE"
 fi
-su - postgres -c "psql postgres -c \"$COMMAND USER $POSTGRES_USER WITH SUPERUSER ENCRYPTED PASSWORD '$POSTGRES_PASS';\""
+su - postgres -c "psql postgres -c \"$COMMAND ROLE $POSTGRES_USER WITH LOGIN SUPERUSER CREATEDB CREATEROLE INHERIT REPLICATION  ENCRYPTED PASSWORD '$POSTGRES_PASS';\""
```
Comment (gubuntu) [inline setup-user.sh]: aren't some of these redundant? i.e a superuser can createdb, createrole etc. 

The replication user must not have other properties, just replication (not superuser)

File: setup-user.sh [hunk 7]
```diff
@@ -19,9 +19,19 @@ echo "postgresql user: $POSTGRES_USER" > /tmp/PGPASSWORD.txt
...
+
+echo "Creating replication user $REPLICATION_USER"
+RESULT_REPLICATION=`su - postgres -c "psql postgres -t -c \"SELECT 1 FROM pg_roles WHERE rolname = '$REPLICATION_USER'\""`
+COMMANDS="ALTER"
+if [ -z "$RESULT_REPLICATION" ]; then
+  COMMANDS="CREATE"
+fi
+su - postgres -c "psql postgres -c \"$COMMANDS USER $REPLICATION_USER WITH REPLICATION ENCRYPTED PASSWORD '$REPLICATION_PASS';\""
+
```
Comment (gubuntu) [inline setup-user.sh:37]: so in a master db, if you don't create a replication user then it can't have any slaves. Is creating a replication user optional?

**Review by gubuntu — COMMENTED** [review]
besides specific code comments, have you tested it?

**Comment by NyakudyaA** [comment]
> besides specific code comments, have you tested it?

Yes, I have tested it and it replicates

File: README.md (hunk 2 above)
Comment (NyakudyaA) [reply README.md]: will change the grammar here

File: README.md (hunk 3 above)
Comment (NyakudyaA) [reply README.md:230]: Without giving select permissions the replication user was failing to replicate anything and the error was `No permission for the required table` Replication is Read-Only on the slave. So any normal user on the slave will not access anything because they are not configured in pg_hba.conf for replication

File: Dockerfile (hunk 1 above)
Comment (NyakudyaA) [reply Dockerfile:40]: They are added here to generate locale for the whole system and on the DB side they are then specified here https://github.com/kartoza/docker-postgis/blob/develop/setup-database.sh#L26
with the env variables controlled here https://github.com/kartoza/docker-postgis/blob/develop/env-data.sh#L101

File: setup-database.sh (hunk 5 above)
Comment (NyakudyaA) [reply setup-database.sh:67]: Postgres users because of the permissions it has.

File: setup-user.sh (hunk 6 above)
Comment (NyakudyaA) [reply setup-user.sh]: I forgot to revert it to the original one will change it to 
`su - postgres -c "psql postgres -c \"$COMMAND USER $POSTGRES_USER WITH SUPERUSER ENCRYPTED PASSWORD '$POSTGRES_PASS';\""`

File: setup-user.sh (hunk 7 above)
Comment (NyakudyaA) [reply setup-user.sh:37]: This is not optional like the DB user. If you do not create a user it will create one for you with the role being `replicator` and password being `replicator`. Because this is needed by pg_hba.conf to enable the replication to actually start

**Comment by NyakudyaA** [comment]
Merging for testing in develop
